libcxx/include/__expected/bad_expected_access.h

```diff
@@ -27,6 +27,8 @@ _LIBCPP_BEGIN_NAMESPACE_STD
 template <class _Err>
 class bad_expected_access;
 
+_LIBCPP_DIAGNOSTIC_PUSH
+_LIBCPP_CLANG_DIAGNOSTIC_IGNORED("-Wweak-vtables")
```

Comment:
Please add a comment why we disabled `-Wweak-vtables` here. We shouldn't just disable it because it happens to be diagnosed. We generally try to avoid weak vtables. FWIW I don't think we should disable it here either. We should instead provide a strong definition in the dylib if it's available on the target.

Reply:
We could. Do you think that would that be worth an ABI break?

Reply:
I don't think we have to make it an ABI break. The weak symbols have, as the name suggests, weak linkage. So we can simply provide a strong definition in the dylib and only enable the generation of weak symbols on platforms which don't have them in the dylib yet. Similar to how we conditionally provide additional instantiations of iostream types.

Reply:
AFAIK We only use weak symbols for the replaceable functions and some debug functions. Not for other functions. I'm not sure whether we want do that. It would allow users to provide strong definitions of destructors they should not provide.

I can't find the example you mean, do you mean the templates guarded by `_LIBCPP_AVAILABILITY_HAS_ADDITIONAL_IOSTREAM_EXPLICIT_INSTANTIATIONS_1` ?

Reply:
> AFAIK We only use weak symbols for the replaceable functions and some debug functions. Not for other functions. I'm not sure whether we want do that. It would allow users to provide strong definitions of destructors they should not provide.

inline symbols have weak linkage, we don't have to give that explicitly. You can see that in https://godbolt.org/z/a3P4P7qhE. The `.weak test()` gives `test()` weak linkage, and explicitly giving it weak linkage doesn't make any difference.

> I can't find the example you mean, do you mean the templates guarded by `_LIBCPP_AVAILABILITY_HAS_ADDITIONAL_IOSTREAM_EXPLICIT_INSTANTIATIONS_1` ?

Yes.

Reply:
This is true for inline non-member functions. However for member functions there is a difference
https://godbolt.org/z/hP9rzfE5o.

I did test with it, but when I put a definition in the header I can't put a second definition in the dylib. Do you have a suggestion how to do this without an ABI break?

Reply:
There is only a difference for special member functions, since there are multiple sometimes. I don't know why clang decides to call the complete object destructor instead of the base object destructor when adding `[[gnu::weak]]` (maybe because it's considered replaceable with `[[gnu::weak]]`), but I don't think that's relevant. The fact that they are weak symbols without `[[gnu::weak]]` doesn't change.

What exactly did you try to add a definition to the dylib?

Reply:
In the header I removed `_LIBCPP_HIDE_FROM_ABI_VIRTUAL` and changed `= default` to `{}`, I also tried with `_LIBCPP_OVERRIDABLE_FUNC_VIS` and `_LIBCPP_WEAK` in the dylib I tried with and without `_LIBCPP_WEAK`. And all permutations of this; but clang dislikes the re-declaration.

Reply:
@philnik777 friendly ping. I really like to move this patch forward since it blocks testing with Clang 19 in the CI.

Reply:
@mordante Is Clang trunk complaining about the vtable of `bad_expected_access<void>` only, or about the one for the base template as well? Cause for the base template we can't do anything, but we could indeed do what @philnik777 said for the `<void>` specialization. We were basically "lazy" when we introduced `bad_expected_access` and decided not to put it in the dylib, but we could have.

Reply:
I checked and only disabling it for the `void` specialization works. However I tried to use `_LIBCPP_WEAK` with that specialization and that resulted in compilation errors. If you or @philnik777 know directly how to solve this feel free to commit patches to this review to solve it.